### PR TITLE
Refactor Philox to expose Counter-based RNG API

### PR DESCRIPTION
**Pull request by jkelling**
This PR refactors alpaka's Philox RNG implementation to expose the counter-based core RNG to the user without overhead.

## Rationale

The current alpaka RNG API is inspired by `<random>` and thereby follows the pattern of classical sequence-based RNGs. The Philox RNG is a counter-based RNG, i.e. it actually is a block cipher that encrypts a value (sequence index/counter) to produce a random number. Thereby, the counter-based RNG can generate any random number from its range for a given seed (key) arbitrarily without following any sequence.

The current API exposes a classical sequence RNG implemented based on the counter-based one by maintaining a counter for each thread while using the grid-thread-index as an offset (value in another word of the global counter) to select independent sequences. The API forces the user to keep an instance of counter state in order to use the RNG. This API limits applications in two ways:
1. There is an overhead for maintaining the counter-state in situations where the index is naturally generated from the simulation step and element each time a random number is needed.
   1. The counter is bundled with the globally-shared key. In many setups having only one global instance of the key/seed, instead of a copy per thread should be more efficient.
2. Relying on the counter will, for the same seed, produce different random numbers depending on `workDiv`, which prohibits reproducibility across platforms. A way to get reproducible runs is to derive the counter/index from the simulation step and element (e.g. time-step and updated cell or particle, this must be some parameter of the simulation continuously progressing, not based on the simulation state.)

## Solution in this PR

1. Split the Philox class hierarchy into the bare counter-based-RNG implementation and the current API and add a new user-facing generator wit default params `PhiloxStateless4x32x10Vector`.
2. Functions related to the core implementation are not using any data members of the Philox classes, hence they are made static.
3. Add a cast operator from `CudaVectorArrayWrapper` to `std::array` in order to allow for seamless transfer of keys and counter offsets between different backends.
4. Add example `counterBasedRng` which fill a multi-dimensional arrays with random numbers in a way yielding the identical results for any backend. This for illustration and constitutes a test by comparing results from one active parallel backend to some form the sequential backend.

## Review

* Please have a look at the proposed API in this PR. I am using this implementation in a current project and thus would appreciate if changes were not too breaking...
* @j-stephan @psychocoderHPC @bernhardmgruber

### CODE REVIEW

File: include/alpaka/rand/Philox/PhiloxBaseCommon.hpp
Comment (jkelling) [inline include/alpaka/rand/Philox/PhiloxBaseCommon.hpp:175]: This code moved, but this comment was also removed, so I feel the need to answer the question posed: The `bumpKey` function constitutes the key schedule of the block cipher, that is for each round a new key is derived from the actual main key. The key schedule does not carry over from block to block, otherwise we would have a stream cipher. The only use in having the round key would be adding additional rounds of encryption, but this would be done by adjusting `numRounds()`, so there is no conceivable reason to return `key`.

**Comment by jkelling** [comment]
Believe @sliwowitz implemented the alpaka Philox, so maybe you should have a look.

File: include/alpaka/rand/Philox/PhiloxBaseCommon.hpp
Comment (sliwowitz) [reply include/alpaka/rand/Philox/PhiloxBaseCommon.hpp:175]: Good, thanks for the explanation. I went by the reference implementation, but I've seen some implementations did bump&return the key at each iteration - I thought those were wring, but wasn't sure about it so I left the comment here as a reminder I wanted to investigate more about it.

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (bernhardmgruber) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
        ALPAKA_FN_HOST_ACC constexpr operator std::array<value_type, size>() const
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (bernhardmgruber) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
            ret[3] = this->w;
            return ret;
        }
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (bernhardmgruber) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
            static_assert(std::tuple_size_v<Other> == size, "Can only convert between vectors of same size.");
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (bernhardmgruber) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
                std::is_same_v<typename Other::value_type, value_type>,
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (bernhardmgruber) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]: Same changes as pointed out above.

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (bernhardmgruber) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]: Same changes as pointed out above.

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (bernhardmgruber) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]: Same changes as pointed out above.

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (bernhardmgruber) [inline example/counterBasedRng/src/counterBasedRng.cpp:86]: Is this really code a user should write? If yes, then we should try to move some parts into alpaka and provide building blocks. If no, we should try to provide an easier example.

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (bernhardmgruber) [inline example/counterBasedRng/src/counterBasedRng.cpp:118]: We cannot build/run this example in the CI with other backends? why?

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (bernhardmgruber) [inline example/counterBasedRng/src/counterBasedRng.cpp]: Please use camel or snake case for local variables.

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]
Is the example enough for a test? I feel like we should have a unit test as well for the philox, or some of its guts.

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (psychocoderHPC) [reply example/counterBasedRng/src/counterBasedRng.cpp:118]: I think this is the problem we have with all examples. The example is using the hard-coded serial accelerator, therefore this acc is required.

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (bernhardmgruber) [inline example/counterBasedRng/src/counterBasedRng.cpp]: Can this be removed?

File: include/alpaka/rand/Philox/PhiloxStateless.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/Philox/PhiloxStateless.hpp]: Is 32bit only a new restriction? Did the previous implementation support 64bit? If no, can we delete the comment?

File: include/alpaka/rand/Philox/PhiloxStateless.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/Philox/PhiloxStateless.hpp]:
```suggestion
            ALPAKA_UNROLL(numRounds())
            for(unsigned int n = 0; n < numRounds(); ++n)
```

File: include/alpaka/rand/Philox/PhiloxStatelessKeyedBase.hpp
Comment (bernhardmgruber) [inline include/alpaka/rand/Philox/PhiloxStatelessKeyedBase.hpp]: Why the r-value reference? Are you missing a move?
```suggestion
        PhiloxStatelessKeyedBase(Key&& key) : m_key(std::move(key))
```

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (jkelling) [reply example/counterBasedRng/src/counterBasedRng.cpp:86]: Yes and No: This is code the user should write only if they want to fill an array with random number, but this is not a common use case. Usually, the user will generate the random number and use them in the same place. Writing it into an array is only a stand-in for mapping random numbers to extrinsic properties of your problem, here this is coordinates.

I do not see part of the RNG-related logic, that could be turned into generic building blocks here.

The only that could be improved would be to have an `operator=` to replace lines 70-72.

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (jkelling) [reply example/counterBasedRng/src/counterBasedRng.cpp]: Indeed.

Comment (jkelling) [inline example/counterBasedRng/src/counterBasedRng.cpp]:
```suggestion
```

File: include/alpaka/rand/Philox/PhiloxStateless.hpp
Comment (jkelling) [reply include/alpaka/rand/Philox/PhiloxStateless.hpp]: The comment was there in the original code, and was only moved. @sliwowitz what do you think?

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (jkelling) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
            ret[0] = this->x;
            return ret;
        }
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (jkelling) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
            ret[1] = this->y;
            return ret;
        }
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (jkelling) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
            ret[2] = this->z;
            return ret;
        }
```

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (jkelling) [inline example/counterBasedRng/src/counterBasedRng.cpp]:
```suggestion
    const int maxPrintFalseResults = extent[2] * 2;

    auto aHost = alpaka::experimental::readAccess(bufHost);
    auto aAcc = alpaka::experimental::readAccess(bufHostDev);
    for(Idx z = 0; z < aHost.extents[0]; ++z)
        for(Idx y = 0; y < aHost.extents[1]; ++y)
            for(Idx x = 0; x < aHost.extents[2]; ++x)
            {
                Data const& valHost(aHost(z, y, x));
                Data const& valAcc(aAcc(z, y, x));
                if(valHost != valAcc)
                {
                    if(falseResults < maxPrintFalseResults)
                        std::cerr << "host[" << z << ", " << y << ", " << x << "] = " << valHost << " != acc[" << z
                                  << ", " << y << ", " << x << "] = " << valAcc << std::endl;
                    ++falseResults;
                }
            }

    if(falseResults == 0)
    {
        std::cout << "Execution results correct!" << std::endl;
        return EXIT_SUCCESS;
    }
    else
    {
        std::cout << "Found " << falseResults << " false results, printed no more than " << maxPrintFalseResults
                  << "\n"
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (jkelling) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
        ALPAKA_FN_HOST_ACC constexpr operator std::array<value_type, size>() const
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (jkelling) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
        ALPAKA_FN_HOST_ACC constexpr operator std::array<value_type, size>() const
```

File: include/alpaka/meta/CudaVectorArrayWrapper.hpp
Comment (jkelling) [inline include/alpaka/meta/CudaVectorArrayWrapper.hpp]:
```suggestion
        ALPAKA_FN_HOST_ACC constexpr operator std::array<value_type, size>() const
```

File: include/alpaka/rand/Philox/PhiloxStateless.hpp
Comment (jkelling) [inline include/alpaka/rand/Philox/PhiloxStateless.hpp:107]:
```suggestion
            ALPAKA_UNROLL(numRounds());
```

**Comment by jkelling** [comment]
> Is the example enough for a test? I feel like we should have a unit test as well for the philox, or some of its guts.

There is a test for RNGs `test/unit/rand/`, I think they use the default RNG for each backend and Philox is the default at least for OpemMP5 and OpenACC. I do not know what exactly it tests for, though. There is also `example/randomStrategies/`, which is hard-coded to use Philox.

File: example/counterBasedRng/src/counterBasedRng.cpp
Comment (jkelling) [reply example/counterBasedRng/src/counterBasedRng.cpp:118]: It is expected to be run with another backend enabled, but the serial one is to serve as the host reference, and thus must also be enabled. Technically, any CPU backend could play this role, but I do not know of a utility we have to select any CPU enabled backend (and fail if none are enabled). `alpaka::ExampleDefaultAcc<Dim, Idx>;` is used as the other, the device, backend.

Comment (bernhardmgruber) [reply example/counterBasedRng/src/counterBasedRng.cpp:118]: Ok, got it!

**Comment by bernhardmgruber** [comment]
I will wait for @sliwowitz to review and approve now.

File: include/alpaka/rand/Philox/PhiloxStateless.hpp
Comment (sliwowitz) [reply include/alpaka/rand/Philox/PhiloxStateless.hpp]: I haven't got around to implementing the 64 bit version - the relevant constants are present, but right now I'm not sure what else would have to be done to make it work. I think we can delete the comment.

**Comment by jkelling** [comment]
@sliwowitz ping?

**Review by sliwowitz — DISMISSED** [review]

**Comment by sliwowitz** [comment]
> @sliwowitz ping?

I like it. 

BTW, tangentially related to this PR - there's been some discussion about moving the RNGs to a separate library,  we might want to revisit the idea in some upcoming meting.

**Review by bernhardmgruber — DISMISSED** [review]

**Comment by bernhardmgruber** [comment]
Can you please rebase your branch on the latest `develop`? Thx!

**Comment by jkelling** [comment]
>  haven't got around to implementing the 64 bit version - the relevant constants are present, but right now I'm not sure what else would have to be done to make it work. I think we can delete the comment.

Done.

**Comment by jkelling** [comment]
> Can you please rebase your branch on the latest `develop`? Thx!

Done. Waiting for CI.

**Review by bernhardmgruber — APPROVED** [review]

**Pull request merged**
